core/src/main/java/org/jdbi/v3/core/config/ConfigRegistry.java

```diff
@@ -68,21 +75,39 @@ private ConfigRegistry(ConfigRegistry that) {
         if (lookup != null) {
             return configClass.cast(lookup);
         }
-        synchronized (createLock) {
+        C config = configClass.cast(configFactory(configClass).apply(this));
+        return Optional.ofNullable(configClass.cast(configs.putIfAbsent(configClass, config))).orElse(config);
+    }
+
+    private Function<ConfigRegistry, JdbiConfig<?>> configFactory(Class<? extends JdbiConfig<?>> configClass) {
+        return configFactories.computeIfAbsent(configClass, klass -> {
+            final Exception ignored;
             try {
-                C config;
-                try {
-                    config = configClass.getDeclaredConstructor(ConfigRegistry.class).newInstance(this);
-                } catch (NoSuchMethodException e) {
-                    config = configClass.getDeclaredConstructor().newInstance();
-                    config.setRegistry(this);
-                }
-                return Optional.ofNullable(configClass.cast(configs.putIfAbsent(configClass, config))).orElse(config);
+                MethodHandle mh = MethodHandles.publicLookup().findConstructor(klass,
+                        MethodType.methodType(void.class, ConfigRegistry.class))
+                    .asType(MethodType.methodType(JdbiConfig.class, ConfigRegistry.class));
+                return Unchecked.function(registry -> (JdbiConfig<?>) mh.invokeExact(registry));
+            } catch (NoSuchMethodException ign) {
```

Comment:
I tend to name the exception for what it signifies rather than what to do with it later, so that the _next_ line reads like "ignore <this thing>". `noRegistryConstructor` or some such in this case. Like how an event handler should call what it does, not what it handles.